src/providers/wfs/qgswfsdescribefeaturetype.cpp

```diff
@@ -14,21 +14,67 @@
  ***************************************************************************/
 
 #include "qgswfsdescribefeaturetype.h"
+#include "qgsmessagelog.h"
 
 QgsWFSDescribeFeatureType::QgsWFSDescribeFeatureType( QgsWFSDataSourceURI &uri )
   : QgsWfsRequest( uri )
 {
 }
 
 bool QgsWFSDescribeFeatureType::requestFeatureType( const QString &WFSVersion,
-    const QString &typeName, bool bUsePlural )
+    const QString &typeName, const QString &namespaces, bool bUsePlural )
 {
   QUrl url( mUri.requestUrl( QStringLiteral( "DescribeFeatureType" ) ) );
   url.addQueryItem( QStringLiteral( "VERSION" ), WFSVersion );
-  url.addQueryItem( bUsePlural ?
-                    QStringLiteral( "TYPENAMES" ) : QStringLiteral( "TYPENAME" ), typeName );
 
-  return sendGET( url, true, false );
+  bool tryNameSpacing;
+  QString prefixOfTypename;
+  if ( ( tryNameSpacing = ( !namespaces.isEmpty() && typeName.contains( ':' ) ) ) )
+    prefixOfTypename = typeName.section( ':', 0, 0 );
+
+  bool success;
+  if ( bUsePlural )
+  {
+    url.addQueryItem( QStringLiteral( "TYPENAMES" ),  typeName );
+    if ( tryNameSpacing )
+    {
+      QString namespacesQueryValue = "xmlns(" + prefixOfTypename + "," + namespaces + ")";
+      url.addQueryItem( QStringLiteral( "NAMESPACES" ), namespacesQueryValue );
+    }
+    success = sendGET( url, true, false );
+  }
+  else
+  {
+    url.addQueryItem( QStringLiteral( "TYPENAME" ),  typeName );
+    if ( tryNameSpacing )
+    {
+      QString namespacesQueryValue = "xmlns(" + prefixOfTypename + "=" + namespaces + ")";
+      url.addQueryItem( QStringLiteral( "NAMESPACE" ), namespacesQueryValue );
+    }
+    success = sendGET( url, true, false );
+  }
+  return success;
+}
+
+bool QgsWFSDescribeFeatureType::requestFeatureType( const QString &WFSVersion,
+    const QString &typeName, const QString &namespaces )
+{
+  bool success = requestFeatureType( WFSVersion, typeName, namespaces, false );
+  if ( !success )
+  {
+    if ( WFSVersion.startsWith( QLatin1String( "2.0" ) ) )
+    {
+      success = requestFeatureType( WFSVersion, typeName, namespaces, true );
```

Comment:
I'm a bit concerned by the systematic retry for the plural form here and below. In case of failures for other reason we delay the error report. On the other hand it is difficult to infer from the error messages when to retry with plural or not.

Reply:
So was I. An idea may be to store on the provider which form to use. Any idea on how that may be done? As of now qgswfsdescribefeaturetype.cpp doesn't know the provider

Reply:
A possibility would be to have in the URI a `usePlural` (or something like that) parameter. Not super user friendly admitedly. When you add a connexion through the GUI, it could try to autodetect which kind of server it is, and automatically add it to layers added in the project.